### PR TITLE
DataFusion 16.0.0 release prep: Update version + add changelog

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/4776

# Rationale for this change

# What changes are included in this PR?

- [x] Update version number
- [x] Generate changelog -- [rendered version](https://github.com/andygrove/arrow-datafusion/blob/datafusion-16/datafusion/CHANGELOG.md)

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by andygrove** [comment]
I am running into GitHub API rate limits trying to generate the changelog. I will try again later today.

**Comment by andygrove** [comment]
The number of request that we make increases over time as we add more issues and PRs. I wonder if we have not hit the limit and need to find a new approach to generating the changelog?

```
When using GITHUB_TOKEN , the rate limit is 1,000 requests per hour per repository. 
```

**Comment by alamb** [comment]
> 
I am running into GitHub API rate limits trying to generate the changelog. I will try again later today.



I used to hit these limits too in arrow-rs -- I spent some time tweaking the various settings for changelog generator -- maybe you can port some of that to datafusion?

https://github.com/apache/arrow-rs/blob/master/.github_changelog_generator

The cache  and max issues thing I believe I remember being important here:
https://github.com/apache/arrow-rs/blob/master/dev/release/update_change_log.sh#L74-L84

**Comment by andygrove** [comment]
Thanks @alamb I will look into that. I guess I just got lucky on my latest attempt and managed to generate it.

File: datafusion/CHANGELOG.md [hunk 1]
```diff
@@ -19,6 +19,337 @@
...
+- Make DataFrame API consuming \(\#4621\) [\#4624](https://github.com/apache/arrow-datafusion/pull/4624) ([tustvold](https://github.com/tustvold))
+- Make execute\_stream functions sync [\#4608](https://github.com/apache/arrow-datafusion/pull/4608) ([tustvold](https://github.com/tustvold))
+- Remove ObjectStore from FileStream \(\#4533\) [\#4601](https://github.com/apache/arrow-datafusion/pull/4601) ([tustvold](https://github.com/tustvold))
+- Remove `AggregateState` wrapper [\#4582](https://github.com/apache/arrow-datafusion/pull/4582) ([alamb](https://github.com/alamb))
+- Fix querying and defining  table / view names with period [\#4530](https://github.com/apache/arrow-datafusion/pull/4530) [[sql](https://github.com/apache/arrow-datafusion/labels/sql)] ([alamb](https://github.com/alamb))
+- refactor code about `subquery_alias` and `expr-alias`. [\#4451](https://github.com/apache/arrow-datafusion/pull/4451) [[sql](https://github.com/apache/arrow-datafusion/labels/sql)] ([jackwener](https://github.com/jackwener))
+
+**Implemented enhancements:**
```
Comment (alamb) [inline datafusion/CHANGELOG.md:48]: this is a pretty epic list

**Review by alamb — APPROVED** [review]
Thank you @andygrove

**Comment by alamb** [comment]
I took the liberty of merging from `apache/master` to this branch to resolve a conflict

**Pull request merged**